### PR TITLE
Tidyup

### PR DESCRIPTION
**Pull request by nitincchauhan**
also fix:
https://docs.oracle.com/javase/7/docs/api/java/sql/Connection.html#close()
states
"Calling the method close on a Connection object that is already closed is a no-op."
but HikariCP throws exception

### CODE REVIEW

**Comment by nitincchauhan** [comment]
CI failure is :

java.lang.NoSuchMethodError: org.hamcrest.Matcher.describeMismatch(Ljava/lang/Object;Lorg/hamcrest/Description;)V
	at com.zaxxer.hikari.metrics.prometheus.HikariCPCollectorTest.connection1(HikariCPCollectorTest.java:54)

File: src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java [hunk 1]
```diff
@@ -455,7 +455,7 @@ private static Connection getClosedConnection()
             public Object invoke(Object proxy, Method method, Object[] args) throws Throwable
             {
                final String methodName = method.getName();
-               if ("abort".equals(methodName)) {
+               if ("close".equals(methodName) || "abort".equals(methodName)) {
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java]: What's going on here?  Multiple-close is already handled -- see [``testDoubleClose()``](https://github.com/brettwooldridge/HikariCP/blob/dev/src/test/java/com/zaxxer/hikari/pool/TestConnections.java#L220).  This proxy only handles methods not intercepted by ``ProxyConnection``.  The ``close()`` method is already intercepted and is pointless here.

File: src/main/java/com/zaxxer/hikari/pool/PoolBase.java [hunk 2]
```diff
@@ -445,9 +439,6 @@ private int getAndSetNetworkTimeout(final Connection connection, final long time
                if (validationTimeout < SECONDS.toMillis(1)) {
                   LOGGER.warn("{} - A validationTimeout of less than 1 second cannot be honored on drivers without setNetworkTimeout() support.", poolName);
                }
-               else if (validationTimeout % SECONDS.toMillis(1) != 0) {
-                  LOGGER.warn("{} - A validationTimeout with fractional second granularity cannot be honored on drivers without setNetworkTimeout() support.", poolName);
-               }
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: Why was this removed?  This is intended to log a warning in cases where the user configures a ``validationTimeout`` with fractional second granularity, for example, ``2250ms`` but does not have a driver that supports ``setNetworkTimeout()``.

File: src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java (hunk 1 above)
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java]: agreed and undone

File: src/main/java/com/zaxxer/hikari/pool/PoolBase.java (hunk 2 above)
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: This is lack of (reading) documentation that leaks in source with checks and messages.

File: src/main/java/com/zaxxer/hikari/pool/PoolBase.java [hunk 3]
```diff
@@ -441,13 +435,7 @@ private int getAndSetNetworkTimeout(final Connection connection, final long time
...
 
-               LOGGER.info("{} - Driver does not support get/set network timeout for connections. ({})", poolName, e.getMessage());
-               if (validationTimeout < SECONDS.toMillis(1)) {
-                  LOGGER.warn("{} - A validationTimeout of less than 1 second cannot be honored on drivers without setNetworkTimeout() support.", poolName);
-               }
-               else if (validationTimeout % SECONDS.toMillis(1) != 0) {
-                  LOGGER.warn("{} - A validationTimeout with fractional second granularity cannot be honored on drivers without setNetworkTimeout() support.", poolName);
-               }
+               LOGGER.info("{} - Failed to get/set network timeout for connection. ({})", poolName, e.getMessage());
             }
```
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: Pool logs failure, e.getMessage() has actual reason.

File: src/main/java/com/zaxxer/hikari/pool/PoolBase.java [hunk 4]
```diff
@@ -519,16 +507,15 @@ private void createNetworkTimeoutExecutor(final DataSource dataSource, final Str
...
-    * @param connectionTimeout the timeout in milliseconds
     */
-   private void setLoginTimeout(final DataSource dataSource, final long connectionTimeout)
+   private void setLoginTimeout(final DataSource dataSource)
    {
       if (connectionTimeout != Integer.MAX_VALUE) {
          try {
-            dataSource.setLoginTimeout((int) MILLISECONDS.toSeconds(Math.max(1000L, connectionTimeout)));
+            dataSource.setLoginTimeout(Math.max(1, (int) MILLISECONDS.toSeconds(500L + connectionTimeout)));
          }
```
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: as for setNetworkTimeout() rounding off to 'nearest' second

File: src/main/java/com/zaxxer/hikari/HikariConfig.java [hunk 5]
```diff
@@ -243,8 +243,8 @@ public long getValidationTimeout()
    @Override
    public void setValidationTimeout(long validationTimeoutMs)
    {
-      if (validationTimeoutMs < 250) {
-         throw new IllegalArgumentException("validationTimeout cannot be less than 250ms");
+      if (validationTimeoutMs < 1000) {
+         throw new IllegalArgumentException("validationTimeout cannot be less than 1000ms");
       }
```
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/HikariConfig.java]: for setNetworkTimeout() as in docs, accepts ms but driver should round off to 'nearest' second.
here safe guarding less than 1 second values, faulty driver may round it to ZERO and treat as 'no timeout'

File: src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java [hunk 6]
```diff
@@ -229,16 +224,15 @@ public final void close() throws SQLException
...
             }
 
             if (dirtyBits != 0) {
                poolEntry.resetConnectionState(this, dirtyBits);
-               lastAccess = clockSource.currentTime();
             }
 
             delegate.clearWarnings();
+            lastAccess = clockSource.currentTime();
          }
```
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java]: setting lastAccess here.. only after 'successful' any sql calls before and lastly clearWarnings() ...closest before recycling

File: src/main/java/com/zaxxer/hikari/pool/HikariPool.java [hunk 7]
```diff
@@ -493,14 +490,14 @@ public void run() {
    private void abortActiveConnections(final ExecutorService assassinExecutor)
    {
       for (PoolEntry poolEntry : connectionBag.values(STATE_IN_USE)) {
+         Connection connection = poolEntry.close();
          try {
```
Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/pool/HikariPool.java]: ...cancels leak task first. this pattern should be followed 'consistently'
proxyConnection.cancelLeakTask() looks like redundent / patchwork and can be removed with refactoring

File: src/main/java/com/zaxxer/hikari/HikariConfig.java (hunk 5 above)
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/HikariConfig.java]: You can't do this, this was specifically changed for a reason:

https://github.com/brettwooldridge/HikariCP/issues/547

Please check the "Blame" history of a file before changing particular lines.  The blame for this line would show this bug specifically.

Comment (nitincchauhan) [inline src/main/java/com/zaxxer/hikari/HikariConfig.java]: that issue was with connectionTimeout. linking it with validationTimeout is wrong.

Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/HikariConfig.java]: That is incorrect, you need to read the entire thread.  Having a ``validationTimeout`` that is higher than the ``connectionTimeout`` leads to violations of the ``connectionTimeout`` contract.  The ``violationTimeout`` absolutely must be able to be set to an equal or lesser value than ``connectionTimeout``.

I am going to cherry-pick the changes in this pull request rather than discussing back and forth on every change.

**Comment by brettwooldridge** [comment]
Cherry-picked.
